### ISSUE TITLE
some confusions about weights function

### ISSUE THREAD

**Issue by SilviaChen1998** [kosukeimai/MatchIt#97]
Hi, thanks for your very useful package.  I want to know how the weights are attributed to each observations.

I find out that "weights[i0]" is calculated by "(treated_by_sub/control_by_sub)[psclass[i0]]", and "control_by_sub <- setNames(tabulateC(psclass[i0], nlevels(psclass)), levels(psclass))". However, I don't know what does the tabulateC function do in this code.
tabulateC <- function(bins, nbins = NULL) {
    .Call(`_MatchIt_tabulateC`, bins, nbins)
}

Could you please describe the way of attributing weights for me, or explain the  "_MatchIt_tabulateC" ? That will be very helpful to me. 
Thanks!!!

**Comment by ngreifer** [comment]
Hello,

Please see the discussion here: #83.

`tabulateC()` is identical to the base R function `tabulate()` but programmed in C++ to be faster. You can actually take a look at the source code [here](https://github.com/kosukeimai/MatchIt/blob/master/src/internal.cpp).

`control_by_sub` is a named vector containing the number of control units in each subclass. For example, it will look something like `c("1" = 100, "2" = 150, "3" = 90, "4" = 90)` which tells you that subclass 1 contains 100 control units, subclass 2 contains 150 control units, etc. `treated_by_sub` is the same but for treated units. So, `(treated_by_sub/control_by_sub)` contains the ratio of the number of treated units to the number of control units in each subclass. Subsetting by `psclass`, which contains subclass membership for each individual, extracts this ratio for each unit; that ratio is used as the weight for each unit. Because this is the formula for ATT weights, only the control units need to have their weights computed; `i0` indexes the control units.

In other words, what this code does is, for each control unit, it finds the ratio of treated units to control units in its subclass. The code implements this in a very fast way, taking advantage of R's vectorization instead of looping through one unit at a time.

**Comment by SilviaChen1998** [comment]
Thanks a lot for your patient explanation. Very clear.

The reason that I was trapped before is due to an explanation in a research paper, in which, the authors used MatchIt to do exact matching and said "Weights are assigned to each observation such that the total weight for both smokers(treated) and nonsmokers(control) in each subclass is equal to one." After reading your comments, I think they might have misunderstood it. It should be "the average weight of every smoker(treated) and nonsmoker(control)" rather than "the total weight for both smokers(treated) and nonsmokers(control) in each subclass" which is equal to one.
